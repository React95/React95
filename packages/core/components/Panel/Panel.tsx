import React from 'react';
import styled from 'styled-components';
import CSS from 'csstype';
import border from './../shared-style/Border';

interface IStyledPanel {
  shadow?: 'intrude' | 'extrude' | 'in-out' | 'none';
  background?: CSS.Property.Background;
  padding?: CSS.Property.Padding;
  direction?: CSS.Property.FlexDirection;
  style?: React.CSSProperties;
  asCanvas?: boolean;
  isScrollable?: boolean;
}
export const StyledPanel = styled.section<IStyledPanel>`
  display: flex;
  flex-grow: 1;
  flex-direction: ${({ direction }) => direction};
  padding: ${({ padding }) => padding};
  overflow-x: ${({ direction }) =>
    direction == 'row' || direction == 'row-reverse' ? 'auto' : 'visible'};
  overflow-y: ${({ direction }) =>
    direction == 'column' || direction == 'column-reverse'
      ? 'auto'
      : 'visible'};
  background-color: ${({ asCanvas, theme, background }) =>
    background
      ? background
      : asCanvas
      ? theme.colors.canvas
      : theme.colors.material};
  color: ${({ background, theme, asCanvas }) =>
    background || asCanvas
      ? theme.colors.canvasText
      : theme.colors.materialText};
  ${({ shadow }) => {
    let shadowStyle;
    if (shadow == 'extrude')
      shadowStyle = border({ direction: 'extrude', useBoxShadow: true });
    else if (shadow == 'intrude')
      shadowStyle = border({ direction: 'intrude', useBoxShadow: true });
    return shadowStyle;
  }}
  &[direction="row"] > & + & {
    border-left: 0;
  }
  &[direction='column'] > & + & {
    border-top: 0;
  }
`;

interface IScrollWrap {
  direction: CSS.Property.FlexDirection;
  padding: CSS.Property.Padding;
}
const ScrollWrap = styled.div<IScrollWrap>`
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  padding: 4px;
  overflow: auto;
  display: flex;
  padding: ${({ padding }) => padding};
  flex-direction: ${({ direction }) => direction};
`;

type ConditonalWrapperProps = {
  children: React.ReactElement;
  condition: boolean;
  wrapper: (children: React.ReactElement) => JSX.Element;
};
const ConditonalWrapper: React.FC<ConditonalWrapperProps> = ({
  condition,
  wrapper,
  children,
}) => (condition ? wrapper(children) : children);

const Panel: React.FunctionComponent<IStyledPanel> = ({
  children,
  shadow = 'intrude',
  background = undefined,
  direction = 'column',
  asCanvas = false,
  style,
  padding = '12px',
  isScrollable = true,
}) => {
  let el;
  if (shadow === 'in-out') {
    el = (
      <StyledPanel shadow="intrude" padding="4px" style={style}>
        <StyledPanel
          asCanvas={asCanvas}
          shadow="extrude"
          background={background}
          direction={direction}
          padding={isScrollable ? 0 : padding}
        >
          <ConditonalWrapper
            condition={isScrollable}
            wrapper={children => (
              <ScrollWrap padding={padding} direction={direction}>
                {children}
              </ScrollWrap>
            )}
          >
            <>{children}</>
          </ConditonalWrapper>
        </StyledPanel>
      </StyledPanel>
    );
  } else {
    // intrude, extrude
    el = (
      <StyledPanel
        asCanvas={asCanvas}
        shadow={shadow}
        background={background}
        direction={direction}
        style={style}
        padding={isScrollable ? 0 : padding}
      >
        <ConditonalWrapper
          condition={isScrollable}
          wrapper={children => (
            <ScrollWrap padding={padding} direction={direction}>
              {children}
            </ScrollWrap>
          )}
        >
          <>{children}</>
        </ConditonalWrapper>
      </StyledPanel>
    );
  }
  return el;
};

Panel.defaultProps = {
  shadow: 'intrude',
  background: undefined,
  direction: 'column',
};

export default Panel;
