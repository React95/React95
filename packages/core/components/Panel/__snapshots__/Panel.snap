// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> Snapshots should match snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
  overflow-x: visible;
  overflow-y: auto;
  background-color: #c6c6c6;
  color: #0a0a0a;
  position: relative;
  border-width: 1px;
  border-style: solid;
  border-top-color: #848584;
  border-right-color: #fefefe;
  border-bottom-color: #fefefe;
  border-left-color: #848584;
  box-shadow: inset -1px -1px 0 0px  #dfdfdf , inset 1px 1px 0 0px  #0a0a0a;
}

.c1[direction="row"] > .c0 + .c0 {
  border-left: 0;
}

.c1[direction='column'] > .c0 + .c0 {
  border-top: 0;
}

.c2 {
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  padding: 4px;
  overflow: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 12px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <section
    class="c0 c1"
    direction="column"
  >
    <div
      class="c2"
      direction="column"
    >
      Dorem ipsum
    </div>
  </section>
</div>
`;
